lineax/_solver/blocktridiagonal.py

```diff
@@ -0,0 +1,125 @@
+from typing import Any
+from typing_extensions import TypeAlias
+
+import jax.lax as lax
+import jax.numpy as jnp
+from jaxtyping import Array, PyTree
+
+from .._operator import (
+    AbstractLinearOperator,
+    blocktridiagonal,
+    is_blocktridiagonal,
+    MatrixLinearOperator,
+)
+from .._solution import RESULTS
+from .._solve import AbstractLinearSolver, linear_solve
+from .misc import (
+    pack_structures,
+    PackedStructures,
+    ravel_vector,
+    transpose_packed_structures,
+    unravel_solution,
+)
+
+
+_BlockTridiagonalState: TypeAlias = tuple[tuple[Array, Array, Array], PackedStructures]
+
+
+class BlockTridiagonal(AbstractLinearSolver[_BlockTridiagonalState]):
+    """Block tridiagonal solver for linear systems, using the Thomas algorithm."""
+
+    def init(self, operator: AbstractLinearOperator, options: dict[str, Any]):
+        del options
+        if operator.in_size() != operator.out_size():
+            raise ValueError(
+                """`BlockTridiagonal` may only be used for linear solves with 
+                square matrices"""
+            )
+        if not is_blocktridiagonal(operator):
+            raise ValueError(
+                """`BlockTridiagonal` may only be used for linear solves with 
+                block tridiagonal `matrices`"""
+            )
+        return blocktridiagonal(operator), pack_structures(operator)
+
+    def compute(
+        self,
+        state: _BlockTridiagonalState,
+        vector,
+        options,
+    ) -> tuple[PyTree[Array], RESULTS, dict[str, Any]]:
+        (diagonal, lower_diagonal, upper_diagonal), packed_structures = state
+        del state, options
+        vector = ravel_vector(vector, packed_structures)
+
+        #
+        # Modifications to basic Thomas algorithm to work on block matrices
+        # notation from: https://en.wikipedia.org/wiki/Tridiagonal_matrix_algorithm
+        # _p indicates prime, ie. `d_p` is the variable name for d' on wikipedia
+        #
+
+        block_size = diagonal.shape[1]
+        size = diagonal.shape[0]
+
+        vector = vector.reshape(size, block_size)
+
+        def blockthomas_scan(prev_cd_carry, bd):
+            c_p, d_p, step = prev_cd_carry
+            # the index of `a` doesn't matter at step 0 as
+            # we won't use it at all. Same for `c` at final step
+            a_index = jnp.where(step > 0, step - 1, 0)
+            c_index = jnp.where(step < size, step, 0)
+
+            b, d = bd
+            a, c = lower_diagonal[a_index, :, :], upper_diagonal[c_index, :, :]
+
+            denom = MatrixLinearOperator(b - jnp.matmul(a, c_p))
+
+            def matrix_linear_solve_vec(step, x_vec):
+                y = linear_solve(denom, x_vec).value
+                step += 1
+                return step, y
+
+            _, new_d_p = matrix_linear_solve_vec(0, d - jnp.matmul(a, d_p))
+            _, new_c_p = lax.scan(matrix_linear_solve_vec, 0, c.T)
```

Comment:
I think this `scan` can be replaced by `vmap` which ends up being a significant speedup on GPU

Reply:
Thanks for reviewing this, I have swapped to vmap in latest commit